src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -48,6 +48,51 @@ internal abstract class IpcEndpoint
         public abstract Task WaitForConnectionAsync(CancellationToken token);
     }
 
+    internal class IpcEndpointHelper
+    {
+        public static Stream Connect(string address, TimeSpan timeout)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                var namedPipe = new NamedPipeClientStream(
+                    ".",
+                    address,
+                    PipeDirection.InOut,
+                    PipeOptions.None,
+                    TokenImpersonationLevel.Impersonation);
+                namedPipe.Connect((int)timeout.TotalMilliseconds);
+                return namedPipe;
+            }
+            else
+            {
+                var socket = new IpcUnixDomainSocket();
+                socket.Connect(new IpcUnixDomainSocketEndPoint(address), timeout);
+                return new ExposedSocketNetworkStream(socket, ownsSocket: true);
+            }
+        }
+
+        public static async Task<Stream> ConnectAsync(string address, CancellationToken token)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
```

Comment:
This was mainly just moving existing code into a shared class, do you suggest we implement this now or do it as a separate PR at a later point? This class is internal, so we are free to continue extending on the transport support going forward and try to reduce the scope of this PR, OK?

Reply:
I'd prefer we resolved it in this PR because it appears like this PR is the one introducing the issue. When this code was in PidIpcEndpoint we were only reasoning about what the default transport was for a given OS and it was pretty closely localized with the code producing the remainder of the address string. Now that the code moved here we are expecting it correctly infer the transport for any address, including non-default ones. This PR also introduced the Parse code which is a new 2nd location that is duplicating the decision making about which transport is implied.

I didn't think this is a significant change. I was expecting this to be ~20-30 lines of code. Am I underestimating what it takes to fix this?

Reply:
I look into it.

Reply:
I wanted to echo Noah's point. Ideally, we'd have a single point where we parse/validate an input address and then internally we just pass around the resolved address which contains all the necessary information to disambiguate. Once we're past the parse/validate point, I'd prefer to not have strings getting passed around.

> PidEndpoint.GetDefaultAddress()

👍

Reply:
I implemented ability to pass around IpcEndpointConfig and use that in these methods, so now you can use an IpcEndpointConfig when creating DiagnosticClient and that will be used through the whole chain all the way down to the Connect/ConnectAsync methods.